networkx/algorithms/dag.py

```diff
@@ -82,172 +84,183 @@ def is_directed_acyclic_graph(G):
+        If G is not a directed acyclic graph (DAG) no topological sort exists
+        and a NetworkXUnfeasible exception is raised.  This can also be
+        raised if G is changed while the returned iterator is being processed.
+
+    RuntimeError
+        If G is changed while the returned iterator is being processed.
+
+    Examples
+    ---------
+    To get the reverse order of the topological sort::
+
+    >>> DG = nx.DiGraph([(1, 2), (2, 3)])
+    >>> list(reversed(list(nx.topological_sort(DG))))
+    [3, 2, 1]
 
     Notes
     -----
     This algorithm is based on a description and proof in
-    The Algorithm Design Manual [1]_ .
+    Introduction to algorithms - a creative approach [1]_ .
 
     See also
     --------
-    is_directed_acyclic_graph
+    is_directed_acyclic_graph, lexicographical_topological_sort
 
     References
     ----------
-    .. [1] Skiena, S. S. The Algorithm Design Manual  (Springer-Verlag, 1998). 
-        http://www.amazon.com/exec/obidos/ASIN/0387948600/ref=ase_thealgorithmrepo/
+    .. [1] Manber, U. (1989). Introduction to algorithms - a creative approach. Addison-Wesley.
+        http://www.amazon.com/Introduction-Algorithms-A-Creative-Approach/dp/0201120372
     """
     if not G.is_directed():
         raise nx.NetworkXError(
             "Topological sort not defined on undirected graphs.")
 
-    # nonrecursive version
-    seen = set()
-    order = []
-    explored = set()
-
-    if nbunch is None:
-        nbunch = G.nodes()
-    for v in nbunch:     # process all vertices in G
-        if v in explored:
-            continue
-        fringe = [v]   # nodes yet to look at
-        while fringe:
-            w = fringe[-1]  # depth first search
-            if w in explored:  # already looked down this branch
-                fringe.pop()
-                continue
-            seen.add(w)     # mark as seen
-            # Check successors for cycles and for new nodes
-            new_nodes = []
-            for n in G[w]:
-                if n not in explored:
-                    if n in seen:  # CYCLE !!
-                        raise nx.NetworkXUnfeasible("Graph contains a cycle at %s." % n)
-                    new_nodes.append(n)
-            if new_nodes:   # Add new_nodes to fringe
-                fringe.extend(new_nodes)
-            else:           # No new nodes so w is fully explored
-                explored.add(w)
-                order.append(w)
-                fringe.pop()    # done considering this node
-    if reverse:
-        return order
-    else:
-        return list(reversed(order))
+    indegree_map = {v: d for v, d in G.in_degree() if d > 0}
+    # These nodes have zero indegree and ready to be returned.
+    zero_indegree = [v for v, d in G.in_degree() if d == 0]
+
+    while zero_indegree:
+        node = zero_indegree.pop()
+        if node not in G:
+            raise RuntimeError("Graph changed during iteration")
+        for _, child in G.edges(node):
```

Comment:
Some algorithms are edge oriented and some are node oriented.

The node oriented algorithms should be able to get neighbors (and succs and preds) without having to worry about edges.  The edge algorithms should be able to get edges.  But these are separate activities.  That's why we have the method successors separate from edges.

Top_sort is (in my view) a node oriented algorithm. Unfortunately we don't have the right methods for predecessors to allow it to work that way.